data/account/participationRegistry.go

```diff
@@ -0,0 +1,947 @@
+
+	log   logging.Logger
+	store db.Pair
+	mutex deadlock.RWMutex
+
+	writeQueue chan partDBWriteRecord
+	asyncErr   error
+
+	flushDone      *sync.Cond
+	flushesPending int
+}
+
+type updatingParticipationRecord struct {
+	ParticipationRecord
+
+	required bool
+}
+
+type partDBWriteRecord struct {
+	insertID ParticipationID
+	insert   Participation
+
+	registerUpdated map[ParticipationID]updatingParticipationRecord
+
+	delete ParticipationID
+
+	flush bool
+}
+
+func (db *participationDB) initializeCache() error {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+
+	db.cache = make(map[ParticipationID]ParticipationRecord)
+	db.dirty = make(map[ParticipationID]struct{})
+
+	records, err := db.getAllFromDB()
+	if err != nil {
+		return err
+	}
+
+	for _, record := range records {
+		// Check if it already exists
+		if _, ok := db.cache[record.ParticipationID]; ok {
+			return ErrMultipleKeysForID
+		}
+		db.cache[record.ParticipationID] = record
+	}
+
+	return nil
+}
+
+func (db *participationDB) writeThread() {
+	needsFlush := false
+	var err error
+	var lastErr error
+	for {
+		var wr partDBWriteRecord
+		var chanOk bool
+		// A call to .Flush() waits until the flash command passes the command queue, and also that the command queue is empty.
+		if needsFlush {
+			select {
+			case wr, chanOk = <-db.writeQueue:
+				if !chanOk {
+					return // chan closed
+				}
+			default:
+				// nothing available, flush completed
+				db.mutex.Lock()
+				needsFlush = false
+				db.asyncErr = lastErr
+				lastErr = nil
+				// see Flush() for the Wait() half of this.
+				db.flushDone.Broadcast()
+				db.mutex.Unlock()
+			}
+		} else {
+			// blocking read until next activity or close
+			wr, chanOk = <-db.writeQueue
+			if !chanOk {
+				return // chan closed
+			}
+		}
+		if len(wr.registerUpdated) != 0 {
+			err = db.registerInner(wr.registerUpdated)
+		} else if !wr.insertID.IsZero() {
+			err = db.insertInner(wr.insert, wr.insertID)
+		} else if !wr.delete.IsZero() {
+			err = db.deleteInner(wr.delete)
+		} else if wr.flush {
+			err = db.flushInner()
+			needsFlush = true
+		}
+		if err != nil {
+			lastErr = err
+		}
+	}
+}
+
+func (db *participationDB) insertInner(record Participation, id ParticipationID) (err error) {
+
+	var rawVRF []byte
+	var rawVoting []byte
+
+	if record.VRF != nil {
+		rawVRF = protocol.Encode(record.VRF)
+	}
+	if record.Voting != nil {
+		voting := record.Voting.Snapshot()
+		rawVoting = protocol.Encode(&voting)
+	}
+
+	err = db.store.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		result, err := tx.Exec(
+			insertKeysetQuery,
+			id[:],
+			record.Parent[:],
+			record.FirstValid,
+			record.LastValid,
+			record.KeyDilution,
+			rawVRF)
+		if err != nil {
+			return fmt.Errorf("unable to insert keyset: %w", err)
+		}
+		rows, err := result.RowsAffected()
+		if err != nil {
+			return fmt.Errorf("unable to insert keyset: %w", err)
+		}
+		if rows != 1 {
+			return fmt.Errorf("unexpected number of rows")
```

Comment:
it's good that you've tested the number of rows - but in case there is an inconsistency, make sure to add the number of actual rows to the error so that we can figure out what went wrong.